src/mpid/ch4/shm/xpmem/xpmem_init.h

```diff
@@ -0,0 +1,155 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef XPMEM_INIT_H_INCLUDED
+#define XPMEM_INIT_H_INCLUDED
+
+#include "xpmem_impl.h"
+#include "xpmem_seg.h"
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_mpi_init_hook
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided,
+                                                       int *tag_bits)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int i, my_local_rank = -1, num_local = 0;
+    xpmem_segid_t *xpmem_segids = NULL;
+    int local_rank_0 = -1;
+    MPIDU_shm_seg_t shm_seg;
+    MPIDU_shm_barrier_t *shm_seg_barrier = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_INIT_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_INIT_HOOK);
+    MPIR_CHKPMEM_DECL(1);
+
+    /* Try to share entire address space */
+    MPIDI_XPMEM_global.segid = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE,
+                                          MPIDI_XPMEM_PERMIT_VALUE);
+    /* 64-bit segment ID or failure(-1) */
+    MPIR_ERR_CHKANDJUMP(MPIDI_XPMEM_global.segid == -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_make");
+    XPMEM_DBG_PRINT("xpmem_init: make segid: 0x%lx\n", (uint64_t) MPIDI_XPMEM_global.segid);
+
+    /* TODO: need a common routine to provide local info before committing comm_world. */
+    for (i = 0; i < size; i++) {
+        MPIDI_av_entry_t *av = NULL;
+        av = MPIDIU_comm_rank_to_av(MPIR_Process.comm_world, i);
+        if (MPIDI_av_is_local(av)) {
+            if (local_rank_0 == -1)
+                local_rank_0 = i;       /* unique ID for temporary shm_seg. */
+
+            if (i == rank)
+                my_local_rank = num_local;
+            num_local++;
+        }
+    }
+    MPIDI_XPMEM_global.num_local = num_local;
+    MPIDI_XPMEM_global.local_rank = my_local_rank;
```

Comment:
I think we have almost exactly the same code in `fbox_init.h`. Maybe it would make sense to move that code into a utility function and store the result in a common location to avoid duplication?

Reply:
Just checked with Ken about the existing common routines for node info. `MPIR_NODEMAP_get_local_info` does exactly the work. I replaced it in xpmem_init. 

However, for `posix/fbox_init`, we also need ranks of all local procs. It is similar to `MPIR_Find_local_and_external`. It will be good to have `MPIR_Find_local` and use it in `posix`. Let's address it in a separate PR.